src/decision/justification_strategy.h

```diff
@@ -0,0 +1,265 @@
+ * input assertions, and one for skolem definitions. The latter is populated
+ * only as skolems appear in assertions. In this example, we have initially:
+ *   input assertions = { (or (P k) Q) }
+ *   relevant skolem definitions =  {}
+ *   SAT context = {}
+ * Say this module returns (P k) as a decision. When this is asserted to
+ * the theory engine, a call to notifyAsserted is made with (P k). The
+ * skolem definition manager will recognize that (P k) contains k and hence
+ * its skolem definition is activated, giving us this state:
+ *   input assertions = { (or (P k) Q) }
+ *   relevant skolem definitions =  { (ite A (= k 1) (= k 2)) }
+ *   SAT context = { (P k) }
+ * We then proceed by satisfying (ite A (= k 1) (= k 2)), e.g. by returning
+ * A as a decision for getNext.
+ *
+ * Notice that the above policy allows us to ignore certain skolem definitions
+ * entirely. For example, if Q were to have been asserted instead of (P k),
+ * then (ite A (= k 1) (= k 2)) would not be added as a relevant skolem
+ * definition, and Q alone would have sufficed to show the input formula
+ * was satisfied.
+ */
+class JustificationStrategy
+{
+ public:
+  /** Constructor */
+  JustificationStrategy(context::Context* c,
+                        context::UserContext* u,
+                        prop::SkolemDefManager* skdm);
+
+  /** Finish initialize */
+  void finishInit(prop::CDCLTSatSolverInterface* ss, prop::CnfStream* cs);
+
+  /** Presolve, called at the beginning of each check-sat call */
+  void presolve();
+
+  /**
+   * Gets the next decision based on the current assertion to satisfy. This
+   * returns undefSatLiteral if there are no more assertions. In this case,
+   * all relevant input assertions are already propositionally satisfied by
+   * the current assignment.
+   *
+   * @return The next SAT literal to decide on.
+   */
+  prop::SatLiteral getNext();
+
+  /**
+   * Are we finished assigning values to literals?
+   *
+   * @return true if and only if all relevant input assertions are already
+   * propositionally satisfied by the current assignment.
+   */
+  bool isDone();
+
+  /**
+   * Notify this class that assertion is an (input) assertion, not corresponding
+   * to a skolem definition.
+   */
+  void addAssertion(TNode assertion);
+  /**
+   * Notify this class that lem is the skolem definition for skolem, which is
+   * a part of the current assertions.
+   */
+  void addSkolemDefinition(TNode lem, TNode skolem);
+  /**
+   * Notify this class that literal n has been asserted. This is triggered when
+   * n is sent to TheoryEngine. This activates skolem definitions for skolems
+   * k that occur in n.
+   */
+  void notifyAsserted(TNode n);
+
+ private:
+  /**
+   * Helper method to insert assertions in `toProcess` to `d_assertions` or
+   * `d_skolemAssertions` based on `useSkolemList`.
+   */
+  void insertToAssertionList(std::vector<TNode>& toProcess, bool useSkolemList);
```

Comment:
Add a comment that indicates that `toProcess` should not be accessed after calling this method because it can contain nodes with reference count 0 and has been modified. Maybe we should also clear it before returning to be extra safe?

Reply:
Sure, I'll clear it.